### PR TITLE
8809 fix

### PR DESCRIPTION
**Pull request by tony1960**
Not a perfect fix but adds stability

### CODE REVIEW

File: rolocController.cpp [hunk 1]
```diff
@@ -97,6 +98,8 @@ void ROLOCcontroller::init()
     QObject::connect(&mDbusHandler,   SIGNAL(requestSetMode(ROLOC::eLINEFINDER_MODE)), this, SLOT(setModeHandler(ROLOC::eLINEFINDER_MODE))      );
 }
 
+
+
 /**
```
Comment (dmacattack) [inline rolocController.cpp:106]:
```suggestion
/**
```

File: rolocController.cpp [hunk 2]
```diff
@@ -155,6 +159,8 @@ void ROLOCcontroller::pollROLOC()
         else
         {
             // roloc is disconnected
+            qDebug() << "roloc has been disconnected";
```
Comment (dmacattack) [inline rolocController.cpp:165]: Lets wrap this in a debug block. If this is on a timer, you dont want to spam the console

File: rolocController.cpp [hunk 3]
```diff
@@ -244,21 +250,87 @@ bool ROLOCcontroller::rolocHardwarePresent()
...
+    bool retValue = false;
+    int count = 0;
+    QQueue<bool>::iterator i;
+
+    for (i = i2cValid.begin(); i != i2cValid.end(); ++i)
+    {
+        test = *i;
+        if (test)
+        {
+            count++;
```
Comment (rabidmonkeyman) [inline rolocController.cpp:324]: i honestly dont follow what the point of this queue full of bools is nor do i really understand why you only are looking for one true. but if youre only looking for a single true value then you can get rid of the count value and set retValue to true and break out of the loop

**Review by rabidmonkeyman — CHANGES_REQUESTED** [review]

File: rolocController.cpp [hunk 4]
```diff
@@ -244,21 +250,87 @@ bool ROLOCcontroller::rolocHardwarePresent()
...
+    if(data != LINEFINDER_I2C_ID)
+    {
+         QTimer::singleShot(2500, [&]()
+         {
+             rolocSetVolume(ROLOC::eVOLUME_OFF);
+             rolocSetParameters(ROLOC::eMODE_GET_SIGNAL_STRENGTH, ROLOC::eFREQ_512HZ_SONDE);
+             data = m_i2cBus.i2c_readWord(mI2cAddr, ROLOC::eCMD_GET_ID);
+         });
+    }
+
```
Comment (dmacattack) [inline rolocController.cpp]: 1) This functionality outside the purpose of this function, this function is called rolocHardwarePresent(). Its purpose it to check if the roloc hardware is present. Thats it. 

2) initializing the volume and signal strength should be in a state machine. And in fact it was. 

3) This logic is **always** setting the volume and mode, if the user wants to set the volume to something else, then it'll get reset in about 3seconds. 

4) you're setting the data variable again (in the lambda). This does not do what you think it does. The lambda is executed outside the scope of the original function, and setting the data has no impact on the rest of the function

for example :
``` c++
void doThing()
{
    int data = 5;
    qDebug() << "data before = " << data;
    QTimer::singleShot(2500, [&]()
    {
         qDebug() << "singleshot done, set data to 6";
         data = 6;
    });

    qDebug() << "data after = " << data;
}

// ------------- this prints -----------------
data before =  5
data after =  5
singleshot done, set data to 6
```
as you can see data doesnt print 6 ever

File: rolocController.cpp [hunk 5]
```diff
@@ -244,21 +250,87 @@ bool ROLOCcontroller::rolocHardwarePresent()
...
     }
 
-    // udpate dbus
-    mDbusHandler.sendPresent(present);
+    if (i2cValid.size() == 7)
+    {
+        present = evalValidity();
+        i2cValid.dequeue();
+    }
+
```
Comment (dmacattack) [inline rolocController.cpp]: What is this doing. you need to explain this. 
Theres a lot that is not clear here. Why is there a queue of bools, why is a queue of bools even needed, what are these magic numbers of 7 and 10?

File: rolocController.cpp [hunk 6]
```diff
@@ -244,21 +250,87 @@ bool ROLOCcontroller::rolocHardwarePresent()
...
     // return the value
     return present;
 }
 
+
+/**
+ * @brief ROLOCcontroller::evalValidity - determine if there are enough true to
+ * consider the interface valid.
+ */
+bool ROLOCcontroller::evalValidity()
```
Comment (dmacattack) [inline rolocController.cpp:315]: it looks like you're trying to implement some type of hysteresis, there are much easier ways of doing this.

**Review by dmacattack — CHANGES_REQUESTED** [review]
consider implementing a hysteresis like this:

``` c++

A hysteresis is just doing a counter x many times to check if your values are changed (rather than a queue)
// create a member variable for the previous value and hysteresis count
int mHystCount;
bool mPrevPresent;

// then in the code, something like this:
bool present = rolocHardwarePresent();
if (present != mPrevPresent)
{
  if (mHystCount == 5 /*wtv max value you determine */ )
  {
     mPrevPresent = present;
  }
  else
  {
    mHysCount++; // the idea is to do several "checks" to ensure the value changes of present. dont flip flop too much
  }
}
else {
   mHystCount =0; // reset the counter
} 
```

File: rolocController.cpp [hunk 7]
```diff
@@ -124,49 +126,73 @@ void ROLOCcontroller::rolocBusy(ROLOC::eSTATE nextState)
...
+              mPrevPresent = present;
+           }
+           else
+           {
+             mBadReadCount++;
+             mPrevPresent = true;
+           }
+         }
+         else {
+            mBadReadCount = 0; // reset
```
Comment (dmacattack) [inline rolocController.cpp:169]: formatting
```suggestion
         else 
         {
            mBadReadCount = 0; // reset
```

File: rolocController.cpp [hunk 8]
```diff
@@ -197,50 +224,62 @@ void ROLOCcontroller::processRolocData()
...
-
-            double mean = getMean(mDepthAccumulator);
-            double stdDev = qSqrt(getVariance(mDepthAccumulator));
-
-            qDebug() << "Mean: " << mean;
-            qDebug() << "Standard Dev.: " << stdDev;
+            // update the sig strength
+            mROLOCsignalStrenth = rolocData;
+#if DBG_BLOCK
+            qDebug() << "sig strength = " << mROLOCsignalStrenth; // TODO verify if is this working ?
```
Comment (dmacattack) [inline rolocController.cpp:234]: Is this still a TODO ? Just curious

File: rolocController.cpp [hunk 9]
```diff
@@ -197,50 +224,62 @@ void ROLOCcontroller::processRolocData()
...
+            mNumSamples++;
 
-            for(quint8 i = 0; i < mDepthAccumulator.count(); i++)
+            if(mNumSamples > N_MULTI_LF_DEPTH_SAMPLE)
             {
-                quint8 isLessThanLowerBound = mDepthAccumulator.at(i) < mean - stdDev * scaleOfElimination;
-                quint8 isGreaterThanUpperBound = mDepthAccumulator.at(i) > mean + stdDev * scaleOfElimination;
-                quint8 isOutOfBounds = isLessThanLowerBound || isGreaterThanUpperBound;
+                QList<quint8> acceptedDepthReadings;
+                float scaleOfElimination = (float)N_MULTI_LF_DEPTH_DELTA;
```
Comment (dmacattack) [inline rolocController.cpp:249]: Lets use a C++ cast;
```suggestion
                float scaleOfElimination = static_cast<float>(N_MULTI_LF_DEPTH_DELTA);
```

File: rolocController.cpp [hunk 10]
```diff
@@ -197,50 +224,62 @@ void ROLOCcontroller::processRolocData()
...
+                        acceptedDepthReadings.append(mDepthAccumulator.at(i));
+                    }
                 }
-            }
 
-            double finalReading = getMean(acceptedDepthReadings);
-            qDebug() << finalReading;
-            mROLOCdepthMeasurement = finalReading;
+                double finalReading = getMean(acceptedDepthReadings);
+                mROLOCdepthMeasurement = finalReading  * 0.0254;
```
Comment (dmacattack) [inline rolocController.cpp:274]: whats this constant ? can we make it a constant in the anon namespace ?

File: rolocController.cpp [hunk 11]
```diff
@@ -197,50 +224,62 @@ void ROLOCcontroller::processRolocData()
...
+                sendDataReport();
+                mNumSamples = 0;
+                mCurrentMode = ROLOC::eMODE_GET_SIGNAL_STRENGTH;
+                qDebug() << "mFrequency  mFrequency-> " << mFrequency;
```
Comment (dmacattack) [inline rolocController.cpp:278]: lol this debug looks redundant
```suggestion
                qDebug() << "mFrequency-> " << mFrequency;
```

File: rolocController.cpp [hunk 12]
```diff
@@ -523,12 +538,41 @@ void ROLOCcontroller::onFreqSetTimerExpired()
...
+/**
+ * @brief ROLOCcontroller::rolocSetDepthMode - set depth mode
+ * @param mode - mode enum
+ */
+void ROLOCcontroller::rolocSetDepthMode(ROLOC::eLINEFINDER_MODE mode)
+{
+    Q_UNUSED(mode);
+    qint16 depthPacket = 0;
+    depthPacket = 0x0200;
+    depthPacket |= mFrequency;
```
Comment (dmacattack) [inline rolocController.cpp:550]: We can simplify this to one line
```suggestion
    qint16 depthPacket = 0x0200 | mFrequency;
```

But what is the 0x0200 ? Can we add a constant for that ? its a magic number at this point

File: rolocController.cpp [hunk 13]
```diff
@@ -523,12 +538,41 @@ void ROLOCcontroller::onFreqSetTimerExpired()
...
+    int status = m_i2cBus.i2c_writeWord(mI2cAddr, ROLOC::eCMD_INFO, depthPacket);
+    if (status == BAD_DATA_READ)
+    {
+#if DBG_BLOCK
+    qDebug() << "write status --> " << status;
+#endif
+    }
+    mCurrentMode = ROLOC::eMODE_GET_DEPTH_MEASUREMENT;
```
Comment (dmacattack) [inline rolocController.cpp:558]: This if block will ALWAYS have the status = -2. So you dont need to DBG_BLOCK it. In fact, I would suggest making it a warning. 

```suggestion
    if (status == BAD_DATA_READ)
    {
#if DBG_BLOCK
    qWarning() << "write status i2c failed ";
#endif
    }
    mCurrentMode = ROLOC::eMODE_GET_DEPTH_MEASUREMENT;
```

File: rolocController.hpp [hunk 14]
```diff
@@ -78,6 +80,8 @@ private slots:
     ROLOC::eLINEFINDER_FREQ mPendingFreq;
     QQueue<bool>     i2cValid;
     int mDisplayRetry = 0;
+    int mBadReadCount = 0;
+    bool mPrevPresent = true;
```
Comment (dmacattack) [inline rolocController.hpp:84]: Can we move all the variable definitions to the ctor initializer list ?
Like this:
``` c++
RolocController::RolocController()
: .. 
, mBadReadCount (0)
, mPrevPresent(true)
```

File: rolocinfopacket.cpp [hunk 15]
```diff
@@ -67,7 +67,7 @@ void RolocInfoPacket::set(unsigned short packet)
         mArrow = ROLOC::eARROW_CENTER;
     }
 
-    // range limit the data
+    // ra limit the data
```
Comment (dmacattack) [inline rolocinfopacket.cpp:70]: looks like an oopsy
```suggestion
    // range limit the data
```

File: rolocController.cpp [hunk 16]
```diff
@@ -197,50 +224,62 @@ void ROLOCcontroller::processRolocData()
...
-                if(!isOutOfBounds)
+                double mean = getMean(mDepthAccumulator);
+                double stdDev = qSqrt(getVariance(mDepthAccumulator));
+#if DBG_BLOCK
+                qDebug() << "*************************Distance Stats*****************";
+                qDebug() << "*************************** bit set  -> " <<  CHECK_BIT(rolocData, 1)  << "*********************";
+                qDebug() << "smaple size: " << mNumSamples;
+                qDebug() << "sample size: " << mNumSamples;
+                qDebug() << "Mean: " << mean;
+                qDebug() << "Standard Dev.: " << stdDev;
```
Comment (dmacattack) [inline rolocController.cpp:259]: you have a duplicate debug here. Lets remove one
```suggestion
                qDebug() << "sample size: " << mNumSamples;
                qDebug() << "Mean: " << mean;
                qDebug() << "Standard Dev.: " << stdDev;
```

File: rolocController.cpp [hunk 17]
```diff
@@ -197,50 +224,62 @@ void ROLOCcontroller::processRolocData()
...
+#endif
+                for(quint8 i = 0; i < mDepthAccumulator.count(); i++)
                 {
-                    acceptedDepthReadings.append(mDepthAccumulator.at(i));
+                    quint8 isLessThanLowerBound = mDepthAccumulator.at(i) < mean - stdDev * scaleOfElimination;
+                    quint8 isGreaterThanUpperBound = mDepthAccumulator.at(i) > mean + stdDev * scaleOfElimination;
+                    quint8 isOutOfBounds = isLessThanLowerBound || isGreaterThanUpperBound;
+
+                    if(!isOutOfBounds)
```
Comment (dmacattack) [inline rolocController.cpp:267]: I dont know what math you're doing here. But can you add parenthesis to ensure your order of operations is clear? 
EG: 
``` c++
mDeptAccumulator.at(i) < ( mean - (stdDev * scaleOfElimination) );
// or
mDeptAccumulator.at(i) < ( mean - stdDev ) * scaleOfElimination );
```
I honestly dont know how the compiler does the order of operations with regards to < > 
so to be on the safe side you can tell it with some parens ?

**Review by dmacattack — CHANGES_REQUESTED** [review]
I just have some simple requests:
* add some constants to make the code more readable for the future devs (0x0200 & 0.0254)
* add parens to the math formulas to ensure what we want is being executed
* fix the member variable intialization in the ctor

**Comment by tony1960** [comment]
Sorry

**Pull request closed without merging**